source/simulator/nullspace.cc

```diff
@@ -81,196 +90,364 @@ namespace aspect
+        const Tensor<1,dim> translation;
 
       public:
+        //Constructor for TensorFunction that takes a Cartesian direction (1,2, or 3)
+        //and creates a translation along that axis 
         Translation(const unsigned int d)
           :
-          direction(d)
+          translation( Point<dim>::unit_vector(d) )
+        {}
+
+        //Constructor for TensorFunction that takes a vector
+        //and creates a translation along that vector 
+        Translation(const Tensor<1,dim> &t)
+          :
+          translation(t)
         {}
 
         virtual Tensor<1,dim> value(const Point<dim> &) const
         {
-          return Point<dim>::unit_vector(direction);
+          return translation;
         }
     };
 
-  }
+    /**
+     * Perform an MPI sum of the entries of a tensor.
+     */
+    template <int rank, int dim, typename Number>
+    inline
+    Tensor<rank,dim,Number>
+    sum (const Tensor<rank,dim,Number> &local,
+         const MPI_Comm &mpi_communicator)
+    {
+      const unsigned int n_entries = local.n_independent_components;
+      Number entries[ n_entries ];
 
-  template <int dim>
-  void Simulator<dim>::setup_nullspace_removal()
-  {
-    if (parameters.nullspace_removal & NullspaceRemoval::linear_momentum)
-      AssertThrow(false, ExcNotImplemented());
+      for (unsigned int i=0; i< n_entries; ++i)
+        entries[i] = local[ local.unrolled_to_component_indices(i) ];
 
-    std::vector<std_cxx1x::shared_ptr<TensorFunction<1,dim> > > funcs;
+      Number global_entries[ n_entries ];
+      dealii::Utilities::MPI::sum( entries, mpi_communicator, global_entries );
 
-    if (parameters.nullspace_removal & NullspaceRemoval::net_rotation)
-      {
-        if (dim==2)
-          funcs.push_back(std_cxx1x::shared_ptr<TensorFunction<1,dim> >(new internal::Rotation<dim>(0)));
-        if (dim==3)
-          for (unsigned int a=0; a<dim; ++a)
-            funcs.push_back(std_cxx1x::shared_ptr<TensorFunction<1,dim> >(new internal::Rotation<dim>(a)));
-      }
+      Tensor<rank,dim,Number> global;
+      for (unsigned int i=0; i< n_entries; ++i)
+        global[ global.unrolled_to_component_indices(i) ] = global_entries[i];
 
-    if (parameters.nullspace_removal & NullspaceRemoval::net_translation_x)
-      funcs.push_back(std_cxx1x::shared_ptr<TensorFunction<1,dim> >(new internal::Translation<dim>(0)));
+      return global;
+    }
 
-    if (parameters.nullspace_removal & NullspaceRemoval::net_translation_y)
-      funcs.push_back(std_cxx1x::shared_ptr<TensorFunction<1,dim> >(new internal::Translation<dim>(1)));
+    /**
+     * Perform an MPI sum of the entries of a symmetric tensor.
+     */
+    template <int rank, int dim, typename Number>
+    inline
+    SymmetricTensor<rank,dim,Number>
+    sum (const SymmetricTensor<rank,dim,Number> &local,
+         const MPI_Comm &mpi_communicator)
+    {
+      const unsigned int n_entries = local.n_independent_components;
+      Number entries[ n_entries ];
```

Comment:
I think the warning/error @tjhei refers to in the general comments should go away if you write this as
```
  const unsigned int n_entries = SymmetricTensor<rank,dim>::n_independent_components;
  Number entries[ n_entries ];
```
or as
```
  const unsigned int n_entries = SymmetricTensor<rank,dim>::n_independent_components;
  Number entries[ SymmetricTensor<rank,dim>::n_independent_components ];
```